### PR TITLE
Check available episode

### PR DESCRIPTION
**Pull request by gborri**
This function check for episode on torrent/nzb and set status on available if found.
Tested only with torrent.
For color code of new episode no preference, if you want to change it its ok.
- [x] if not curShow.paused chenge check
- [x] move log inside markAvailable procedure

### CODE REVIEW

**Comment by fernandog** [comment]
@gborri can you explain this please? Didn't get it. Why do we want an available status? when it's available SR will snatch it and set as SNATCHED

**Comment by gborri** [comment]
@fernandog for me it's useful and i think for other. The point is that when you add a show know how many episode is available on provider is an information, at least for me, in order to choose which show to start watching. I'm used to start watching a season when all the episode of the season is available. with this function you don't need to look for it on provider and is available on sickrage.

**Comment by gborri** [comment]
@fernandog i forgot to make it configurable, i will try.

**Comment by gborri** [comment]
@fernandog done. Make available check as an option

**Comment by fernandog** [comment]
@gborri is there any DB change?

**Comment by gborri** [comment]
@fernandog only for my db version. SickRage native db will not be touched

**Comment by abeloin** [comment]
Looking at the code. I think we should change AVAILABLE_CHECK to ep_availability_check, as the other term is to generic and isn't self explanatory.

**Comment by abeloin** [comment]
We are migrating TPB? class MigrateSickBeardDB(AlterTVShowsFieldTypes)

**Comment by fernandog** [comment]
@gborri sorry, I don't like this ideia/FR. IMO I wouldn't merge it. Misses the point of SR automation

@abeloin @TagForce what do you think?

File: sickbeard/tv.py [hunk 1]
```diff
@@ -1226,15 +1226,20 @@ def wantEpisode(self, season, episode, quality, manualSearch=False, downCurQuali
...
+            if epStatus in (SKIPPED, IGNORED, ARCHIVED) and not manualSearch:
+                logger.log(u"Existing episode status is skipped/ignored/archived, ignoring found episode", logger.DEBUG)
+                return False
 
         curStatus, curQuality = Quality.splitCompositeStatus(epStatus)
 
         # if it's one of these then we want it as long as it's in our allowed initial qualities
         if quality in anyQualities + bestQualities:
-            if epStatus in (WANTED, UNAIRED, SKIPPED):
+            if epStatus in (WANTED, UNAIRED, SKIPPED) or epStatus in Quality.AVAILABLE:
```
Comment (abeloin) [inline sickbeard/tv.py]: This should be if epStatus in (WANTED, UNAIRED, SKIPPED, AVAILABLE)

**Comment by gborri** [comment]
@fernandog it's not a matter or automation. the point is to know what have been available on torrent provider. For italian language the show has been released lot time after the original release. So there no reason to login on torrent provider to know if it's available or not and doing this everyday/everyweek etc. The point is to let sickrage discover the available torrent file on provider. I have on sickrage 50 Show and looking on provider if episode exit is a waste of time. The FR is enabled only if someone want it, otherwise nothing has been changed. I don't understand why you don't want it to be merged. I think that could be useful for lot of people.

**Comment by gborri** [comment]
@abeloin the code is correct, available status is subjected to quality as snatch and downloaded.

**Comment by gborri** [comment]
@abeloin the migration is need to change my db to be equal to the one of sickrage. Is needed to me and to the few people that have downloaded my repo.

**Comment by gborri** [comment]
@abeloin for the change of name is fine, i will do it.

**Comment by gborri** [comment]
@abeloin have changed AVAILABLE_CHECK into EP_AVAILABILITY_CHECK

File: sickbeard/search.py [hunk 2]
```diff
@@ -1,717 +1,745 @@
...
-
-    origThreadName = threading.currentThread().name
-    threads = []
-
-    show_list = sickbeard.showList
-    fromDate = datetime.date.fromordinal(1)
-    episodes = []
-
-    for curShow in show_list:
-        if not curShow.paused:
```
Comment (abeloin) [inline sickbeard/search.py]: @gborri Why remove this?

**Comment by abeloin** [comment]
@gborri Why do you create a new quality? Can you explain

**Comment by abeloin** [comment]
@gborri Why do you create a new quality?

**Comment by abeloin** [comment]
@gborri Do we need to be informed of all the marking in the log, this generate an excessive amount of log?

**Comment by gborri** [comment]
@abeloin your question is right. I have to remove it but only if EP_AVAILABILITY_CHECK is enabled.

**Comment by gborri** [comment]
@abeloin i haven't got you question about marking. Are you meaning the possibility to see in log file when the status of the episode will be changed?

**Comment by abeloin** [comment]
I'm talking about these 2 lines:
* https://github.com/gborri/SickRage/blob/branch-check_available_episode/sickbeard/search_queue.py#L169
* https://github.com/gborri/SickRage/blob/branch-check_available_episode/sickbeard/search_queue.py#L277

Which double the number of line received: one from the provider and one from this patch. Instead of blindly saying we marked the episode can the log line go inside markAvailable and make sure the episode isn't already marked available before logging us to death with every release of the same episode.

**Comment by gborri** [comment]
@abeloin ok, got it.

**Comment by gborri** [comment]
@abeloin we need to do it also for the snatchEpisode?

**Comment by abeloin** [comment]
@gborri The log line for snatch only happen when we snatch something? If so, it only generate a log line on a snatch event which shouldn't be often

**Comment by gborri** [comment]
@abeloin also snatch could fail and the log has been already writed.

**Comment by gborri** [comment]
@abeloin before merge this PR you need to merge #1344. otherwise we got a lot of KeyError: <sickbeard.tv.TVEpisode object at 0xa91e45cc>

**Comment by abeloin** [comment]
@gborri How come?

**Comment by gborri** [comment]
@abeloin sorry, i haven't understood your question...

**Comment by abeloin** [comment]
@gborri How come this PR is dependent on PR #1344 ?

**Comment by gborri** [comment]
is not dependant, but the this PR will arise a lot of KeyError:
<sickbeard.tv.TVEpisode object at 0xac4e9eac>.
The 1344 will fix it.

**Comment by abeloin** [comment]
@gborri Please explain why you would get the keyerror <sickbeard.tv.TVEpisode object at 0xac4e9eac> and what/where PR1344 fix it.

**Comment by abeloin** [comment]
@gborri Small question: How this patch handle stale result? Do you discard results after some time as elapsed?

**Comment by gborri** [comment]
@abeloin i'm not sure to have got your question. but I will try to respond. 
The status change is permanent. If I found a result on torrent I will expect that persist on torrent.
Some change on show edit start set episode to skipped and restart discovery. for example quality change.

**Comment by gborri** [comment]
@abeloin with Fix #1777 this PR is mergeable

**Comment by gborri** [comment]
@abeloin @fernandog do you think this PR could be merged?

**Comment by fernandog** [comment]
not now. Too mamy issues. We need to make master stable again

**Comment by abeloin** [comment]
@gborri I also wanted to talk with you about the implementation. Looking at this, I was thinking that maybe we could push it a little bit further and offer the choice of the release to download(in case of multiple one). This would require a db for it and a reworking of the statuses and maybe use notification icon when episode are available. Also a management page would be nice.

I'm a little bit busy until the 30th, but after that I think we should talk about it as it could be a good idea

**Comment by WebSpider** [comment]
This effectively starts bothering providers and indexers with API calls for all eps of a series which are set to WANTED or SKIPPED. Now WANTED I see no problem in: Since we want them, we should do our best to find the eps.

SKIPPED essentially means to me "Weeeel, I might be interested in them somewhere in the future. Or not. Don't know". Should we bother providers with querying them for that kind of stuff? SR must start running into API limits for users with huge DB's then.

**Comment by gborri** [comment]
@webspider i think that if you are worried about bothering indexer probably we need to not let show update happens daily. The check four availability is made with same frequency of backloggin (less frequent). I think is a useful feature in order to know which show (episode actually) is available. SR start to check for availability on the day of air and stops when found(every 30 days at most) accordingly to backlogging frequency.

**Comment by WebSpider** [comment]
@gborri we are overhauling the way the indexer works in the show-updater-changes branch, to lower the load on indexers. Once this work is done, I will look at this PR again.

**Comment by gborri** [comment]
@miigotu is there any reason why have you closed this PR?

**Comment by miigotu** [comment]
1. All admin have been skeptical as to the usefulness of this PR from the start
2. If the episode is wanted, it will be snatched, otherwise we cannot make requests to the providers for episodes we might not even want.

**Comment by gborri** [comment]
@miigotu i can rebase with the development branch (i need some help with the new mako template).
The impact to the providers is not more relevant as the as-is solution: it's one query more than today. infact if you set episode as wanted it keep searching on provider till it find it.
The availability check for most of the case make search by season (1 query for all episode) because you look for all episode at once, only for ongoing season the search for episode is made by singolar episode. When the episode is marked available the search would be stopped. I don't know which is your criteria in order to choose which series start to watch, mine is to see if there is at least one season of episode available. Without this feature you need to go to provider manually and check series by series: it's a waste of time.

**Comment by miigotu** [comment]
@gborri we simply do not want this feature because it makes no sense to us. It's ok for us to say no sometimes.

**Comment by gborri** [comment]
@miigotu the real point is that if i need to continuosly merge my repo with sickrage one i will have less time to contribute to the project. Then i have started to port my feature to SickRage because someone said to me that all the feature present on my repo would be ported on sickrage (someone said echelon was interested on it). This function is optional and can activated only if someone is interested otherwise the behaviour is the same as today. Now i'm facing the situation in which the admin (i don't know the reason) don't want to merge this one but all of you has accepted all other feature and i'm in the situation to continuosly need to merge my repo with the sickrage wasting time. Please take in consideration that also the little time the people is investing in this project is precious. Let me know.

**Comment by WebSpider** [comment]
@gborri thank you for putting your time in SR, it is much appreciated.

For this reason exactly, it is best practice to have feature-branches. Put one feature on a branch, so it can be merged into other repositories as needed.

I'd suggest you put this feature in a separate branch, and merge this in your branch, while we may merge this later, if at all. That way, there's maximum flexibility, with minimum overhead.

**Comment by miigotu** [comment]
It takes literally seconds to merge changes from one repo to another, without PR.

**Comment by gborri** [comment]
@miigotu it's not true.you need to manage conflict. in the last case you have changed the template definition and before merging the change i have to understand the new way...... i't simple only if there is no conflict.

**Comment by miigotu** [comment]
Either way you slice it, this PR is not getting added here. When your code is buggy you don't fix it, and most of it has very limited use cases.

Believe me I have spent more time trying to fix one bug your code introduced than it took you to merge back to your branch a hundred times.

**Comment by gborri** [comment]
@miigotu which is the bug you have spent time on? is not true that i don't manage bug i have introduced.

**Comment by gborri** [comment]
@WebSpider i have some doubt that the work that i have done on this project "is much appreciated": has been taken the code i have done that like to admin group and discarded the one that the admin group don't like. Has been said that the feature of my repo were interested and i have merged it into sickrage (with not so easy effort) and now the admin group said that are not interested anymore on it (some of them were also removed). if this was the scenario when i started i would never started. I think that a comunity project will success if the priority and the feature needed are defined by the comunity itself otherwise if are only the admins that can decide which is the evolution of the project i think it will fail as soon as arise an alternative. When i started i though that was a comunity on which contribute to, but from what I have seen today it think that it's a closed group that grab idea from the comunity defining priority on their interest.

**Comment by WebSpider** [comment]
@gborri looking at your contributions that made it to the project, I would disagree when you say it is not appreciated.

It is true that not all code will be merged into mainline, for various reasons, few of them mentioned above.

This is a community project, and will continue to be. Not only do we have over 150 suggestions from users that would like to see features added (and unfortunately insufficient resources to make them all true), we also have various pieces of code that have been submitted without request that have been merged, and several pieces of code that have been rejected, including pieces of code of myself, and other regular committers.

There are reasons not to except code, since all code also has to be maintained. If we see there would be future issues with maintaining code, or if estimations are that this code would introduce a feature for an estimated small amount of users, it may be rejected.

In such cases, it is more productive to make sure people you know who would like this feature :+1: it, instead of having big discussions about it, or code around the objections that were raised by other committers.

**Comment by miigotu** [comment]
1. He said that SOME of your features were welcome.
2. Still why is trakt code you handled adding back shows after they are removed in SR?
3. "if the priority and the feature needed are defined by the comunity itself" It is. You alone are not the community, you are just a small piece of it. In general, our vast majority want easy to use software that works exactly as it is supposed to all of the time, without bugs. Obscure and unused features are not their interest or priority.

If my management style does not suit you then it is ok if you don't contribute, but it is obvious that my approach works or we would not have doubled our user base since I started nor would SR be even remotely as stable as it now is becoming.

In programming it is not about what you can do, it is about what you SHOULD do. We all can code 1000 features based on fleeting ideas that no one will use and add it to the code, but should we? Absolutely not. We add things that are highly requested and will be used by the majority of users. And if we add something that turns out is rarely used or doesn't work as well as we hoped, we remove it.
We also do not add things if they are not received well when proposing them, which is exactly what a PR is, a proposal, a REQUEST. Please don't be butt-hurt just because some requests are denied, that is just life.

Programming is a practice of catering to the majority of your user base and potential users, not the 1%.

**Comment by gborri** [comment]
@miigotu @WebSpider if you really want to follow your style of management why not using a feature for voting feature and issue? like this one http://feathub.com/?

**Comment by miigotu** [comment]
@gborri If it was a much wanted feature, there would be an assload of users on this thread screaming in outrage begging for it. There are not.

**Comment by WebSpider** [comment]
Also, with some 200 fr's out there, feathub may not be a bad idea.

**Pull request closed without merging**